src/pcre2_jit_compile.c

```diff
@@ -13202,22 +13210,28 @@ SLJIT_ASSERT(common->early_fail_start_ptr <= common->early_fail_end_ptr);
 if (private_data_size > 65536)
   {
   SLJIT_FREE(common->private_data_ptrs, allocator_data);
-  SLJIT_FREE(common->optimized_cbracket, allocator_data);
   return PCRE2_ERROR_NOMEMORY;
   }
 
 if (common->has_then)
   {
-  common->then_offsets = (sljit_u8 *)(common->private_data_ptrs + total_length);
+  total_length = ccend - common->start;
+  common->then_offsets = (sljit_u8 *)SLJIT_MALLOC(total_length, allocator_data);
+  if (!common->then_offsets)
+    {
+    SLJIT_FREE(common->private_data_ptrs, allocator_data);
+    return PCRE2_ERROR_NOMEMORY;
+    }
   memset(common->then_offsets, 0, total_length);
   set_then_offsets(common, common->start, NULL);
   }
 
 compiler = sljit_create_compiler(allocator_data);
 if (!compiler)
   {
-  SLJIT_FREE(common->optimized_cbracket, allocator_data);
   SLJIT_FREE(common->private_data_ptrs, allocator_data);
+  if (common->has_then)
+    SLJIT_FREE(common->then_offsets, allocator_data);
   return PCRE2_ERROR_NOMEMORY;
```

Comment:
With all these repeated FREE statements, it's common to have a block at the end of the function with the cleanup code, which you can jump to with goto. Saves duplicating the cleanup. The cleanup block can have multiple jump targets, with fallthrough, so it cleans up stuff in reverse order to what was allocated. Or it can just check each pointer for NULL and clean up all the ones that have been alloced.

Reply:
JIT does not use goto at the moment. But it is reasonable here. I expect compilers to do CSE. But you cannot trust in all compilers. Not sure what is the best. Shall we introduce goto?

Reply:
I don't mind; if you don't want to use goto, then duplicating the FREE calls is OK. It's your code - it's only a maintenance concern.